### PR TITLE
Add glow to object properties to customize lighting

### PR DESCRIPTION
**Pull request by basicer**
As an added bonus setting the color on the entity will also work when this is on.

### CODE REVIEW

**Comment by rubenwardy** [comment]
Looks useful, but it looks like you're just adding a property without ever using it?

File: src/content_cao.h [hunk 1]
```diff
@@ -106,6 +106,7 @@ class GenericCAO : public ClientActiveObject
 	float m_step_distance_counter = 0.0f;
 	u8 m_last_light = 255;
 	bool m_is_visible = false;
+	bool m_fullbright = true;
```
Comment (SmallJoker) [inline src/content_cao.h]: Wrong default value

File: src/content_cao.cpp [hunk 2]
```diff
@@ -660,6 +660,7 @@ void GenericCAO::updateLight(u8 light_at_pos)
 void GenericCAO::updateLightNoCheck(u8 light_at_pos)
 {
 	u8 li = decode_light(light_at_pos);
+	if ( m_fullbright ) return;
```
Comment (SmallJoker) [inline src/content_cao.cpp]: http://dev.minetest.net/Code_style_guidelines
```C++
	if (m_fullbright)
		return;
```

File: src/content_cao.cpp [hunk 3]
```diff
@@ -659,6 +659,9 @@ void GenericCAO::updateLight(u8 light_at_pos)
 
 void GenericCAO::updateLightNoCheck(u8 light_at_pos)
 {
+	if ( m_fullbright )
```
Comment (nerzhul) [inline src/content_cao.cpp]: wrong parenthetis spaces

File: src/object_properties.cpp [hunk 4]
```diff
@@ -58,6 +58,7 @@ std::string ObjectProperties::dump()
 	os<<", makes_footstep_sound="<<makes_footstep_sound;
 	os<<", automatic_rotate="<<automatic_rotate;
 	os<<", backface_culling="<<backface_culling;
+	os<<", fullbright="<<fullbright;
```
Comment (nerzhul) [inline src/object_properties.cpp]: please fix spaces

**Comment by paramat** [comment]
I'm wondering how this could be used, i'm thinking to create mobs that glow in the dark. Do you have any other uses in mind?

I don't feel on/off fullbright is very useful, as often you would want a light level lower than fullbright, because at night fullbright will be blindingly bright, too bright for a ghost mob for example or many other uses.
How about making this a 'minimum light level' integer instead? It can be set to 15 for fullbright, or, if set at a lower level it will cause a mob to glow gently in the dark but also have the correct light level in daytime.

File: doc/lua_api.txt [hunk 5]
```diff
@@ -4107,6 +4107,7 @@ Definition tables
     --  ^ Limit automatic rotation to this value in degrees per second,
     --    value < 0 no limit.
         backface_culling = true, -- false to disable backface_culling for model
+        fullbright = false, -- disable lighting
```
Comment (paramat) [inline doc/lua_api.txt]: 'disable lighting' bad description, this implies a dark object.

**Comment by paramat** [comment]
Does a new object property need adding to network protocol? If so i forgot to add 'can zoom'.

File: src/network/networkprotocol.h [hunk 6]
```diff
@@ -178,6 +178,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 			Breaks compatibility with older clients.
 	PROTOCOL VERSION 36:
 		Backwards compatibility drop
+		Add fullbright to object properties
```
Comment (paramat) [inline src/network/networkprotocol.h]: Not sure this needs adding here.

**Comment by basicer** [comment]
The intention is basicly to tell the engine not to mess with the color in regard tl lighing and let me manage that myself.

For entities like the world edit position marker I want this to always be Fulbright always even when around another node. (in world edit they make the entity texture black to work around this but with this patch they could change to something more reasonable).

For something like a glowing mob I could sample the light level myself and calculate the color that makes the entity look how is like.  Doing this I can still adjust my color on the fly instead of having it clobbered to grayscale by the lighting system.

Having a max light level of 15 would still clobber the color if you set it to 15.

The Fulbright code in this patch does fix the blackness when an entity is in the ground, I'm already using it.

File: src/network/networkprotocol.h (hunk 6 above)
Comment (nerzhul) [reply src/network/networkprotocol.h]: it should as it's transported on network

Comment (paramat) [reply src/network/networkprotocol.h]: Ok will add my 'can zoom' then.

**Comment by paramat** [comment]
I see, then my title edit was wrong, it really does disable lighting, sorry, reversing.

**Comment by paramat** [comment]
> The Fulbright code in this patch does fix the blackness when an entity is in the ground, I'm already using it.

Well i mean, it makes it fullbright, not the correct ambient light level. So i assume you are using a mod to sample the correct ambient light level and you are using that.

I'm just wondering how we can make this extra property useful for those whp are not using mods as you are. We do not have an object 'glow' property and that would be useful, so is there a way that can also be combined into this PR? How about '-1' disables while 0-15 sets a minimum light level?

**Comment by basicer** [comment]
I was thinking of the -1 solution myself but I was worried it might seem unclean.  Any ideas for a property name? does glow fit in place of min_light?

File: src/script/common/c_converter.cpp [hunk 7]
```diff
@@ -427,6 +427,19 @@ bool getintfield(lua_State *L, int table,
 }
 
 bool getintfield(lua_State *L, int table,
+		const char *fieldname, s8 &result)
+{
+	lua_getfield(L, table, fieldname);
+	bool got = false;
+	if(lua_isnumber(L, -1)){
```
Comment (SmallJoker) [inline src/script/common/c_converter.cpp]: Space missing after `if` and before `{`, otherwise :+1:

**Comment by paramat** [comment]
Thanks, 'glow' is fine. We do have parameters elsewhere which use a negative value to select a special option, so it seems ok to me to do this.

File: src/script/common/c_converter.cpp (hunk 7 above)
Comment (basicer) [reply src/script/common/c_converter.cpp]: This code is copied from the function above with the last parameter type changed.  I'm still new so not sure which is preferred:

a) Follow the style of the surrounding code
b) Follow the style guide introducing inconstancy 
c) Reformat the rest of the file

Comment (paramat) [reply src/script/common/c_converter.cpp]: b)
Follow the code style guidelines but only for the lines of code you are working on or nearby lines.
It's a balance, a) would mean nothing gets corrected, and c) would make PRs too big and complex to review.
A lot of our code still has bad code style, we're slowly correcting it as we work on MT.

**Comment by paramat** [comment]
I'm just wondering what type of 'glow' is more useful:
Minimum light level: light = MAX(ambient, parameter)
or
Ambient plus value: light = MIN(ambient + parameter, max_light)
?
The PR is currently the 2nd option, which makes the object brighter according to ambient light, so that it is always a set amount brighter than ambient light, this might look a little weird.
The 1st option means glow in the dark but matches ambient light in brighter environments.
I'm unsure, any input?

**Comment by basicer** [comment]
I had tried option 1 first, but watching some entities though a few day night cycles it felt sort of strange.  As the world gets darker it actually feels like the entity is getting brighter as night comes, sort of like a nightlight or something.  With glow it just seemed like the entity had some internal light, and felt a bit more natural.

**Comment by paramat** [comment]
It's ok as it is then, it's also consistent with particles.

File: src/script/common/c_content.cpp [hunk 8]
```diff
@@ -275,6 +275,7 @@ void read_object_properties(lua_State *L, int index,
 	}
 	lua_pop(L, 1);
 	getboolfield(L, -1, "backface_culling", prop->backface_culling);
+	getintfield(L, -1, "glow", prop->glow);
```
Comment (paramat) [inline src/script/common/c_content.cpp]: I think you may need to also add this to `push_object_properties()` L302 below.

**Comment by paramat** [comment]
screenshot_20170913_064831

Tested and seems ok.
+1 once my line comment is attended to.

File: src/script/common/c_content.cpp (hunk 8 above)
Comment (paramat) [reply src/script/common/c_content.cpp]: I did this when i added the 'can_zoom' property, see https://github.com/minetest/minetest/commit/561a01cc2a8ca0b83b37975a048d34dcfd0d41e1

Comment (paramat) [reply src/script/common/c_content.cpp]: Yes it's needed for:
```
// get_properties(self)
int ObjectRef::l_get_properties(lua_State *L)
{
	NO_MAP_LOCK_REQUIRED;
	ObjectRef *ref = checkobject(L, 1);
	ServerActiveObject *co = getobject(ref);
	if (co == NULL)
		return 0;
	ObjectProperties *prop = co->accessObjectProperties();
	if (!prop)
		return 0;
	push_object_properties(L, prop);
	return 1;
}
```

File: doc/lua_api.txt [hunk 9]
```diff
@@ -4125,6 +4125,9 @@ Definition tables
     --  ^ Limit automatic rotation to this value in degrees per second,
     --    value < 0 no limit.
         backface_culling = true, -- false to disable backface_culling for model
+        glow = 0,
+    --  ^ Add this much extra lighting when calculating texture color.
+          value < 0 disables light's effect on texture color
```
Comment (paramat) [inline doc/lua_api.txt]: Maybe add a very short explanation of why this is useful and what it allows? (mod setting of color or whatever).

**Comment by paramat** [comment]
:+1:

**Comment by paramat** [comment]
a9d43a04713dd0342d675f3bbb5cbacd863b4118

**Pull request closed without merging**